Lombiq.Tests.UI/Services/FrontendServer.cs

```diff
@@ -124,11 +125,18 @@ public void Configure(
                 Task = cliTask,
                 StopAsync = async () =>
                 {
-                    // This cancellation token forcefully closes the frontend server (i.e. SIGTERM, Ctrl+C), which is
-                    // the only way to shut down most of these servers anyway. For this reason there is no need to await
-                    // the task, and trying to do so would throw OperationCanceledException.
-                    await cancellationTokenSource.CancelAsync();
-                    cancellationTokenSource.Dispose();
+                    // Attempt to close the process with an interrupt signal (SIGINT, same as hitting Ctrl+C), which is
+                    // the only way to shut down most of these servers as they have an infinite main loop. If that
+                    // fails within a minute, the process is killed forcefully (SIGTERM).
+                    await gracefulCancellation.CancelAsync();
+                    forcefulCancellation.CancelAfter(TimeSpan.FromMinutes(1));
+
+                    // Using Task.WhenAny() without unwrapping its output avoids the OperationCanceledException, which
+                    // is not relevant in disposal code.
+                    await Task.WhenAny(cliTask);
+
+                    gracefulCancellation.Dispose();
+                    forcefulCancellation.Dispose();
 
                     await context.PortLeaseManager.StopLeaseAsync(frontendPort);
```

Comment:
Shouldn't these three lines be in a `finally`?

Reply:
`await Task.WhenAny(cliTask);` can, no?

Reply:
Nope, as explained by the comment above it. `await await Task.WhenAny(cliTask);` or `await Task.WhenAny(cliTask).Unwrap();` could throw. `await Task.WhenAny(cliTask);` just finishes with success when an inner task ends.

Reply:
OK then.